common/common.go

```diff
@@ -33,3 +39,33 @@ func (r Role) String() string {
 	}
 	return "standby"
 }
+
+// Write file to temp and atomically move when everything else succeeds.
+// This function is taken from
+//   https://github.com/youtube/vitess/blob/master/go/ioutil2/ioutil.go
+// Copyright 2012, Google Inc. BSD-license
+func WriteFileAtomic(filename string, data []byte, perm os.FileMode) error {
+	dir, name := path.Split(filename)
+	f, err := ioutil.TempFile(dir, name)
+	if err != nil {
```

Comment:
I think upon receiving an error from `TempFile`, it is safe to assume the file was not created (as `TempFile` is implemented in terms of `CreatFile` - whose result it delegates back to the caller).